### PR TITLE
crystal-api : handle exceptions & bump crystal-lang version

### PR DESCRIPTION
**Pull request by MattMacD78**
# What's new?

- updated `shards`
  - to enable compiling binary 
  - to respect `crystal-lang` standard scaffolding _(as expected by `Heroku` for instance)_
  
- split `Dockerfile`
  - to use lighter `alpine`-based images
  - to compile crystal-binary for non-dev `docker-compose` / `Dockerfile`
  - to use `ENTRYPOINT`
    - can still be overridden in `docker-compose-dev` through `${ENTRYPOINT}`
    - ```ENTRYPOINT="sleep infinity" docker-compose -f docker-compose-dev.yml up --build```

- bumped `crystal-lang` **_version_** to **_1.3.0_**

- added print `DEBUG` option

- updated `docker-compose` files to reflect / use all of the above

- updated readme

- added test script

- **_Fixed exceptions_**
  - by rescuing json parsing of 3rd party API response body
  - by adding **_quick & very very dirty handling of rate-limiting_** by 3rd party API 
    - Enrich response w/ `X-retried` header _(`bool`)_ to indicate at least one retry occured
    - `FFU #1` : implement **_retries w/ exponential backoff strategy_**, see https://github.com/feifanzhou/robust_http.cr
    - `FFU #2` : do not even rely on a 3rd party API, implement `GeoIP2` natively, see https://github.com/delef/geoip2.cr

### CODE REVIEW

**Comment by MattMacD78** [comment]
@tdimnet : this is the PR w/ proposed improvements to https://github.com/DataDog/apps/pull/90

File: examples/geomap/README.md [hunk 1]
```diff
@@ -1,7 +1,11 @@
-# Datadog Redis App
+# Datadog Geomap App
```
Comment (tdimnet) [inline examples/geomap/README.md:1]: 👍

File: examples/geomap/README.md [hunk 2]
```diff
@@ -12,8 +16,8 @@
 ## Technologies
 
 - Pigeon Map: https://pigeon-maps.js.org/docs/installation
+- crystal-lang: https://crystal-lang.org
 
 ## Getting Started
 
 @TODO
```
Comment (tdimnet) [inline examples/geomap/README.md:23]: I am going to add the whole description of the project by the end of the week, then I'll create the PR.

File: examples/geomap/crystal-api/.gitignore [hunk 3]
```diff
@@ -0,0 +1,2 @@
+# Binary
+bin
```
Comment (tdimnet) [inline examples/geomap/crystal-api/.gitignore:2]: FIY, I added this: we don't need the binary on the repo

File: examples/geomap/crystal-api/main.cr [hunk 4]
```diff
@@ -49,10 +50,31 @@ server = HTTP::Server.new do |context|
...
+    ## better : implement retries w/ exponential backoff startegy, see https://github.com/feifanzhou/robust_http.cr
+    ## even better : do not rely on 3rd party API, implement GeoIP2 natively, see https://github.com/delef/geoip2.cr
 
     if ! geo.nil?
-        geo_j = JSON.parse(geo.body)
-        geo_h = geo_j.as_h
+        begin
+            geo_j = JSON.parse(geo.body)
+            geo_h = geo_j.as_h
+        rescue
```
Comment (tdimnet) [inline examples/geomap/crystal-api/main.cr:72]: What does it mean? I mean the `rescue` keyword.

File: examples/geomap/docker-compose-dev.yml [hunk 5]
```diff
@@ -1,12 +1,15 @@
...
-    build: crystal-api 
+    build:
+      context: crystal-api
+      target: base
     stdin_open: true
     tty: true
     environment:
       BIND: 0.0.0.0
       PORT: 8080
+      DEBUG: "${DEBUG:-true}"
```
Comment (tdimnet) [inline examples/geomap/docker-compose-dev.yml:12]: You cannot write true? I am not sure I understand why you do this.

**Comment by MattMacD78** [comment]
LGTM

**Pull request merged**